src/main/java/com/mastercard/developer/oauth/OAuth.java

```diff
@@ -171,12 +172,25 @@ static String toOauthParamString(SortedMap<String, List<String>> queryParamsMap,
    * Generates a random string for replay protection as per
    * https://tools.ietf.org/html/rfc5849#section-3.3
    *
-   * @return concatenation of 3 characters from the most significant bits and 13 characters from the least significant bits without dashes.
+   * @return concatenation of character length of {@link OAuth#NONCE_LENGTH} minus least significant bits length from the most significant bits and characters from the least significant bits without dashes.
+   * If the most length + least length less than {@link OAuth#NONCE_LENGTH}, the number of "0" will be appended as a prefix to ensure the nonce length is 16.
    */
   static String getNonce() {
     UUID uuid = UUID.randomUUID();
     String least = Long.toString(uuid.getLeastSignificantBits(), Character.MAX_RADIX).replace("-", "");
-    String most = Long.toString(uuid.getMostSignificantBits(), Character.MAX_RADIX).replace("-", "").substring(0, 3);
+    String most = Long.toString(uuid.getMostSignificantBits(), Character.MAX_RADIX).replace("-", "");
+
+    int remainingLength = NONCE_LENGTH - least.length();
+    int mostLength = most.length();
+
+    if (mostLength >= remainingLength) {
+      most = most.substring(0, remainingLength);
+    } else {
+      for (int i = 0; i < remainingLength - mostLength; i++) {
+        most = "0" + most;
+      }
+    }
+
```

Comment:
since we were having nonce length issues, maybe you could add unit tests that generate 20 (or more) nonces and check their length. I don't think this will take long to run, but if it does, reduce that number (20) to something more sensible.

Reply:
The issue happens **randomly** and **rarely**.
The number of test running doesn't guarantee the test is passed or failed as it's only reproducible by keeping on running the junit test till failed as Jean-Alexis has suggested. So I am bit reluctant to rely on the unit test for this case.

Reply:
oh ok - I thought the bug would be reproducible more often.

Reply:
Guys,

What about `UUID.randomUUID().toString().replace("-", "").substring(0, 16)`?

IMO it's good enough for generating those nonces.

J.-A.

Reply:
Initially, I thought of this simple solution. But I found that the nonce string can be duplicated (quite rare occurrence), so I came up with current solution.

Reply:
Did you experience that yourself? Because randomUUID() is supposed to use a CSPRNG and generate type 4 UUIDs.

Reply:
type 4 UUIDS have 2^128 possible combinations, trimming it down to 16 chars brings it down to 2^64.

if everyone is ok with that kind of clash probability (I think it is ok), then we can use the simpler approach as JA proposed.

Reply:
Also keep in mind that the nonces only have to be unique within the 30 minute sliding window around the current timestamp which makes the clash probabilities infinitely smaller still.

Reply:
Also, if the goal is to shorten the UUID, it would be useful to understand randomness implications of using least/most significant bits and cutting out parts. The only reason I originally used a UUID as a nonce was because the JVM calls a natively compiled implementation which is very fast. Instead of butchering it like this, perhaps it makes more sense to pick a simpler random generation algorithm.

Reply:
@jaaufauvre: I went through the code of UUID and SecureRandom and googled to confirm the duplicate possibility with substring of 16.

@kkrauth: I thought of the 30-minute nonce validity before. In our context, as of now the number of incoming requests are quite average, so it would consider that the chance of duplicate is quite small or zero.
The approach of least and most is to use SecureRandom to create random 8-byte least and 8-byte. Then generating the readable character ([0-0a-z]) from this random least and most. So this approach can be considered similar to create the random 16-character string by looping from 0..15 with SecureRandom generation.

Reply:
I think we can use a much simpler approach to generate a random string of 16 lengths. Java has a SecureRandom class which is cryptographically random number generator. 

The current approach is not similar to SecureRandom because UUID itself is not super random in its entirety because as you can see from the [spec](https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)), around 7 of the most significant bits are fixed to identify the version used and variant. 

You can have a look at this solution using SecureRandom: https://stackoverflow.com/a/157202.

Reply:
Yes, in our case something like `Long.toHexString(Math.abs(secureRandom.nextLong()))` could work.

Reply:
@manixate, @jaaufauvre the UUID.randomUUID is using the SecureRandom, so I don't want to reinvent the wheel on this. But it's up to you guys to decide.
```java
public static UUID randomUUID() {
        SecureRandom ng = Holder.numberGenerator;

        byte[] randomBytes = new byte[16];
        ng.nextBytes(randomBytes);
        randomBytes[6]  &= 0x0f;  /* clear version        */
        randomBytes[6]  |= 0x40;  /* set to version 4     */
        randomBytes[8]  &= 0x3f;  /* clear variant        */
        randomBytes[8]  |= 0x80;  /* set to IETF variant  */
        return new UUID(randomBytes);
    }
```

Reply:
This will not result in a 16 length string every time. `nextLong()` can be anything from 1-32 bits and `toHexString` doesn't guarantee lengths. We can use something like:

```java
static final String CHARS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

String randomString(int len) {
   SecureRandom rnd = new SecureRandom();
   StringBuilder sb = new StringBuilder(len);
   for( int i = 0; i < len; i++ ) {
      sb.append(CHARS.charAt( rnd.nextInt(CHARS.length())));
   }
   return sb.toString();
}
```

Reply:
@manixate Yeah, I didn't test that :)

`String.format("%016x", Math.abs(new SecureRandom().nextLong()))` would be better.

> `nextLong()` can be anything from 1-32 bits

It should be fine I guess, because 0000000000000001 is a legit nonce value.

I'm also OK with the approach above.

Reply:
@manixate, @jaaufauvre: will do the above approach.
Before updating the code, I would like to highlight the solution that I have thought and implemented.
- The 8-byte least and most value are generated with 8-byte filled by SecureRandom.nextBytes more complicated than SecureRandom.nextInt. Please refer to SecureRandom.nextBytes
- The Long.toString(number, MAX_RADIX) is using the radix including [0-9a-z] like your approach
- Why UUID doesn't use the simple approach for random number like suggested solution?

Reply:
@huy1912 Sorry for the late response!

This discussion was interesting. As a summary, I would say:
* Asking for 128-bits of randomness to keep only 64 bits in the end is probably overkill
* Truncating UUID is error prone (we have to understand the UUID spec version we rely on)
* We shouldn't be too paranoid about the randomness. We are not generating cryptographic keys here and the chance of colliding 64-bits values in a 30 minutes window is more than low
* We should use something simple we can port to other languages

Unfortunately I think there are still as many possibilities as there are developers, and you're the one pushing this improvement, so I trust you one this :)

J.-A.